File: backend/models/Event.js
const mongoose = require('mongoose');
const objectId = mongoose.Schema.Types.ObjectId;


const eventSchema = mongoose.Schema({
	name: { type: String, required: true },
	creator: { type: objectId, ref: 'User', required: true },
	status: { type: String, required: true},
	place: { type: String },
	geoLoc: {
		lat: { type: Number },
		long: { type : Number }
	},
	beginDate: { type: Date },
	endDate: { type: Date },
	publicFlag: { type: Boolean, required: true, default: true }, //public keyword is reserved for system
	votingPrerequisites: { type: Boolean, required: true, default: false },
	musicalStyle: { type: String, required: true, default: 'none'},
	physicalEvent: { type: Boolean, required: true, default: false },
	guestsInfo: [{
		userId: { type: objectId, ref: 'User', required: true },
		right: { type: String, required: true}
	}],
	guestsNumber: { type: Number, required: true, default: 0 },
	tracksInfo: [
		{
			trackId: { type: String, required: true },
			trackDuration: {type: Number, required: true, default: 0},
			timeBeginListening: {type: Date, required: true, default: 0},
			votesNumber: {type: Number, required: true }
		}
	]
});

module.exports = mongoose.model('Event', eventSchema);

File: frontend/musicroomfortytwo/Model/Events/Event.swift
//
//  Event.swift
//  musicroomfortytwo
//
//  Created by Tristan Leveque on 05/03/2021.
//

import Foundation
import UIKit

struct TrackInfo {
    var id: String?
    var name: String?
    var artist: String?
    var nbVotes: Int?
    var didUserVote: Bool?
    var timeBeginListening: String?
    var uri: String?
    var duration_ms: Int?
    var coverUrl: String?
    var coverImg: UIImage?
}

class Event {
    var id: String?
    var name: String?
    var creator: String?
    var `public`: Bool?
    var status: String?
    var votingPrerequisites: Bool?
    var musicalStyle: String?
    var guestsNumber: Int?
    var guestsInfo: [GuestInfos]?
    var tracks: [TrackInfo] = []
    var physicalEvent: Bool?
    var place: String?
    var geoLoc: Geoloc?
    var beginDate: String?
    var endDate: String?
    var coverEvent: UIImage?
    
    init(){}
    init(event: EventObject) {
        self.id = event._id
        self.name = event.name
        self.creator = event.creator
        self.public = event.publicFlag
        self.status = event.status
        self.votingPrerequisites = event.votingPrerequisites
        self.musicalStyle = event.musicalStyle
        self.guestsNumber = event.guestsNumber
        self.guestsInfo = event.guestsInfo
        self.physicalEvent = event.physicalEvent
        self.place = event.place
        self.geoLoc = event.geoLoc
        self.beginDate = event.beginDate
        self.endDate = event.endDate
        for i in 0..<event.tracksInfo.count {
            self.tracks.append(TrackInfo(id: event.tracksInfo[i].trackId, nbVotes: event.tracksInfo[i].votesNumber, didUserVote: event.tracksInfo[i].userVote, timeBeginListening: event.tracksInfo[i].timeBeginListening))
        }
    }
    func update(event: EventObject) {
        self.id = event._id
        self.name = event.name
        self.creator = event.creator
        self.public = event.publicFlag
        self.status = event.status
        self.votingPrerequisites = event.votingPrerequisites
        self.musicalStyle = event.musicalStyle
        self.guestsNumber = event.guestsNumber
        self.guestsInfo = event.guestsInfo
        self.physicalEvent = event.physicalEvent
        self.place = event.place
        self.geoLoc = event.geoLoc
        self.beginDate = event.beginDate
        self.endDate = event.endDate
        self.tracks.removeAll()
        for i in 0..<event.tracksInfo.count {
            self.tracks.append(TrackInfo(id: event.tracksInfo[i].trackId, nbVotes: event.tracksInfo[i].votesNumber, didUserVote: event.tracksInfo[i].userVote, timeBeginListening: event.tracksInfo[i].timeBeginListening))
        }
    }
}
